### PR TITLE
Create SQL Extensibility and Gem builder docs

### PR DESCRIPTION
**Pull request by alexanderahn**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207981709272274

### CODE REVIEW

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        # generate the actual macro call given the component's state
+        resolved_macro_name = f"{self.projectName}.{self.name}"
+        non_empty_param = ",".join([param for param in [props.parameter1] if param != ''])
+        return f'{{{{ {resolved_macro_name}({non_empty_param}) }}}}'
+
+```
+
+### Load and Unload Properties
+
+The `loadProperties` and `unloadProperties` methods load and convert the component's state to the default macro property representation.
```
Comment (aarjay00) [inline docs/SQL/extensibility/gem-builder/gem-builder.md:229]: When prophecy parses a macro invocation, it represents a macro definition in a default state: MacroProperties which consists of
 - macro name
 - project name
 - parameters used.


For example if macro invocation is 
```dbt_utils.deduplicate(relation, partition_by, order_by)```

 prophecy parses it into an object:

```MacroProperties(project_name = "dbt_utils", macro_name = "deduplicate", parameters = List(
MacroParameter(value="relation"),
MacroParameter(value="partition_by"),
MacroParameter(value="relation")
)) ```


This object now has to be converted into the gem state defined by user. This logic is defined in loadProperties. 

Similarly the the opposite case where this enhanced UX is not available due to some reason, prophecy needs to be able to render the default macro UI. For this purpose user needs to define logic to convert the gem properties back to the default macro properties object which prophecy understands.

File: docs/package-hub/package-builder/gem-builder.md [hunk 2]
```diff
@@ -563,5 +563,5 @@ We learned the essential code for all Transform Gems using the Limit Gem as an e
 
 **How does Gembuilder work for projects aside from Python and Scala?** Package Hub supports re-usable components in SQL, for example, to create a Macro to use custom Jinja logic. A user can create a Custom Gem to be used within the same project using Macros. As of Prophecy Version 3.2, support for reusable components in SQL is in “beta” release. These are the limitations:
 
-1. A SQL project cannot be marked as package. so it wont show on package hub.
-2. User cannot reuse a model, Gem etc from one project to another for SQL projects.
+1. A SQL project cannot be marked as package. Therefore, it won't show on package hub.
+2. You cannot reuse a model or Gem from one project to another for SQL projects.
```
Comment (aarjay00) [inline docs/package-hub/package-builder/gem-builder.md:567]: You can do it. Not via the package hub. But user can release a project and add it to another projects dependencies via the dependencies setting.

Screenshot 2024-10-23 at 9 32 23 AM

**Review by aarjay00 — CHANGES_REQUESTED** [review]

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+:::
+
+Each Prophecy Model is composed of individual operations, or [Gems](/docs/concepts/project/gems.md), that perform actions on data. While Prophecy offers dozens of Gems out-of-the-box, some data practitioners want to extend this idea and create their own Gems. Gem builder allows enterprise users to add custom Gems. You can create custom source, target, and transformation Gems, and then publish them for your team to use.
+
+Our SQL Gem builder supports Databricks and Snowflake SQL. It's built on dbt Core™, allowing you to build upon existing dbt libraries to define new macros to use in your custom Gem.
+
+You can create a Gem that writes a reference to either of the following options:
+
+- a new user-defined macro or SQL query
```
[the remark this answers is not in the document]
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:22]: Done.

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+
+   ![Gem builder new Gem](img/gem-builder-new.png)
+
+2. Enter a Gem Name, choose a Category, and verify the Directory Path. Then click **Create**. The Gem is automatically set to save in `macros/` as `gem_name.py` and `gem_name.sql` files.
+
+Now you can customize the Gem using the split-screen code editor. See the following [Creating a Gem](#creating-a-gem) section to learn how to define your Gem.
+
+:::note
+
+Custom Gem logic can be shared with other users within the Team and Organization. Navigate to the Gem listing to review Prophecy-defined and User-defined Gems. When your Gem is ready, publish it so that it is available to use in other Models.
```
[the remark this answers is not in the document]
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:39]: Done.

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+### Gem components
+
+There are two types of Gems that you can create while using the Gem builder:
+
+- **DataSource Gems**: These Gems enable the reading and writing of data from or to various data sources.
+- **Transform Gems**: These Gems apply transformations/joins/any other custom logic onto any source that is passed into them.
+
+Programmatically, a Gem is a component with the following parts:
+
+- The **Gem UI Component** which gets user information from the screen. This code is rendered on the Prophecy UI.
```
[the remark this answers is not in the document]
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:56]: Done.

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+There are various UI components that can be defined for custom Gems such as scroll boxes, tabs, and buttons. These UI components can be grouped together in various types of panels to create a custom user experience when using the Gem.
+
+After the Dialog object is defined, it's serialized as JSON, sent to the UI, and rendered there.
+
+Depending on what kind of Gem is being created, a `Dialog` needs to be defined.
+
+#### Column selector
+
+You can use the column selector property if you want to select the columns from UI and then highlight the used columns using the `onChange` function.
```
[the remark this answers is not in the document]
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:184]: Done.

File: docs/SQL/extensibility/gem-builder/gem-builder.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+    def onChange(self, context: SqlContext, oldState: Component, newState: Component) -> Component:
+        # Handle changes in the component's state and return the new state
+        return newState
+
+```
+
+### Apply
+
+The code for the Gem logic is defined in the `apply` function. Here the above User Defined properties are accessible using `self.projectName.{self.name}`.
```
[the remark this answers is not in the document]
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:215]: Done.

File: docs/SQL/extensibility/gem-builder/gem-builder.md (hunk 1 above)
Comment (alexanderahn) [reply docs/SQL/extensibility/gem-builder/gem-builder.md:229]: Done.

File: docs/package-hub/package-builder/gem-builder.md (hunk 2 above)
Comment (alexanderahn) [reply docs/package-hub/package-builder/gem-builder.md:567]: Done.

**Pull request merged**